src/System.Management.Automation/engine/parser/Parser.cs

```diff
@@ -7678,27 +7690,96 @@ private ExpressionAst MemberAccessRule(ExpressionAst targetExpr, Token operatorT
                 member = GetSingleCommandArgument(CommandArgumentContext.CommandArgument) ??
                     new ErrorExpressionAst(ExtentOf(targetExpr, operatorToken));
             }
-            else
+            // Member name may be an incomplete token like `$a.$(Command-Name`; do not look for generic args or
+            // invocation token(s) if the member name token is recognisably incomplete.
+            else if (_ungotToken == null)
             {
+                genericTypeArguments = GenericMethodTypeArgumentsRule(out rBracket);
                 Token lParen = NextInvokeMemberToken();
+
                 if (lParen != null)
                 {
+                    // Fall through to the last available token expected in a member name for calculating the expected
+                    // position of the lParen. This accounts for things like `[Array]::Empty[int[]()` so that we don't
+                    // get unhandled behaviour when input is expected but incomplete.
+                    int endColumnNumber = rBracket?.Extent?.EndColumnNumber
+                        ?? genericTypeArguments?.LastOrDefault()?.Extent?.EndColumnNumber
+                        ?? member.Extent.EndColumnNumber;
+
                     Diagnostics.Assert(lParen.Kind == TokenKind.LParen || lParen.Kind == TokenKind.LCurly, "token kind incorrect");
-                    Diagnostics.Assert(member.Extent.EndColumnNumber == lParen.Extent.StartColumnNumber,
-                                       "member and paren must be adjacent");
-                    return MemberInvokeRule(targetExpr, lParen, operatorToken, member);
+                    Diagnostics.Assert(
+                        endColumnNumber == lParen.Extent.StartColumnNumber,
+                        "member and paren must be adjacent when the method is not generic");
+                    return MemberInvokeRule(targetExpr, lParen, operatorToken, member, genericTypeArguments);
                 }
             }
 
             return new MemberExpressionAst(
-                    ExtentOf(targetExpr, member),
-                    targetExpr,
-                    member,
-                    @static: operatorToken.Kind == TokenKind.ColonColon,
-                    nullConditional: operatorToken.Kind == TokenKind.QuestionDot);
+                ExtentOf(targetExpr, member),
+                targetExpr,
+                member,
+                @static: operatorToken.Kind == TokenKind.ColonColon,
+                nullConditional: operatorToken.Kind == TokenKind.QuestionDot,
+                genericTypeArguments);
+        }
+
+        private List<ITypeName> GenericMethodTypeArgumentsRule(out Token rBracketToken)
+        {
+            List<ITypeName> genericTypes = null;
+
+            int resyncIndex = _tokenizer.GetRestorePoint();
+            Token lBracket = NextToken();
+            rBracketToken = null;
+
+            if (lBracket.Kind == TokenKind.LBracket)
```

Comment:
If you invert this condition, we can dispense with an indentation level. If we call `NextToken()` and the next token is not an LBracket, it sounds like we might need to return the token and indicate that this method does not apply?

Reply:
Returning null just leaves the `MemberAccessRule` with a null value for method type arguments. Since it's impossible to determine the difference between a property and a method in the parser until we get the final `()` token at the end, calling the method here and getting null is OK. The `MemberAccessRule` only needs to know if there were generic method arguments or not and to pass that information along accordingly.

If there were none, a null result is expected and just means the `MemberAccessRule` and the resulting AST element has no generic type arguments. 

Good point on the indent level, though, we can avoid needing to `Resync` here by inverting the conditional and using `PeekToken()` as well which is a nice plus! 👍

Reply:
I lied, we can't avoid a Resync() unfortunately. It is still simpler to code the way you suggested though.

Left a detailed comment in the code explaining why PeekToken() can't be used in that specific place without breaking everything.